File: lib/go/thrift/inbound_stream.go
package thrift

import (
	"fmt"
	"io"
)

type TInboundStream interface {
	io.Closer

	Receive(Context, TRequest) error
}

type tInboundStream struct {
	tBaseStream

	messageType TMessageType
}

func newTClientInboundStream(name string, seqID int32, in, out TProtocol, cl *TSyncClient) *tInboundStream {
	return &tInboundStream{
		tBaseStream: newTClientBaseStream(name, seqID, in, out, SERVER_STREAM_GOAWAY, cl),
		messageType: SERVER_STREAM_MESSAGE,
	}
}

func newTServerInboundStream(name string, seqID int32, in, out TProtocol) *tInboundStream {
	return &tInboundStream{
		tBaseStream: newTServerBaseStream(name, seqID, in, out, CLIENT_STREAM_GOAWAY),
		messageType: CLIENT_STREAM_MESSAGE,
	}
}

func (s *tInboundStream) Close() error {
	defer s.close()

	var err error

	s.goAwayOnce.Do(func() {
		err = s.writeGoAway()

		if err != nil {
			return
		}

		err = s.readGoAwayACK()
	})

	return nil
}

func (s *tInboundStream) readGoAwayACK() error {
	mt, err := s.readShell()

	if err != nil {
		return err
	}

	if mt != s.goAwayACKType {
		return fmt.Errorf("invalid go away ack")
	}

	return nil
}

func (s *tInboundStream) Receive(ctx Context, req TRequest) error {
	var typeID, err = s.readMessageBegin(ctx)

	if err != nil {
		return parseStreamingError(err)
	}

	switch typeID {
	case s.messageType:
		defer s.in.ReadMessageEnd()
		return req.Read(s.in)
	case s.goAwayType:
		s.in.ReadMessageEnd()
		s.writeGoAwayACK()
		s.goAwayOnce.Do(func() {})
		s.close()
		return io.EOF
	default:
		s.in.ReadMessageEnd()
		return fmt.Errorf("unexpected messaege type: %v", typeID)
	}
}

File: lib/rb/lib/thrift/inbound_stream.rb
module Thrift
  class TInboundStream < TBaseStream
    def initialize(iprot, oprot, name, seqid, klass, message_type, unlock = nil)
      @message_type = message_type
      @klass = klass
      @unlock = unlock
      @closing = false
      super(iprot, oprot, name, seqid)
    end

    def close
      wait_ready

      @mutex.synchronize do
        return nil if @closed

        begin
          @closing = true
          write_shell_unlocked @message_type + 1

          loop do
            mtype = read_message_begin

            case mtype
            when @message_type
              @iprot.skip(Types::STRUCT)
              @iprot.read_message_end
            when @message_type + 2
              @iprot.read_message_end
              return
            else
              raise ApplicationException.new(
                ApplicationException::INVALID_MESSAGE_TYPE,
                'invalid message type'
              )
            end
          end
        ensure
          @unlock&.call
        end
      end
    end

    def receive
      wait_ready

      @mutex.synchronize do
        case read_message_begin
        when @message_type
          res = @klass.new
          res.read(@iprot)
          @iprot.read_message_end

          return res.arg
        when @message_type + 1
          @iprot.read_message_end
          write_shell_unlocked @message_type + 2 unless @closing
          @closed = true
          @cond.broadcast
          @unlock&.call

          raise EOFError
        else
          raise ApplicationException.new(
            ApplicationException::INVALID_MESSAGE_TYPE,
            'invalid message type'
          )
        end
      end
    end
  end
end
